azure-iot-device/tests/iothub/pipeline/test_mqtt_topic_iothub.py

```diff
@@ -0,0 +1,1014 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+import pytest
+import logging
+import datetime
+from azure.iot.device.iothub.pipeline import mqtt_topic_iothub
+from azure.iot.device import Message
+
+logging.basicConfig(level=logging.DEBUG)
+
+# NOTE: All tests (that require it) are parametrized with multiple values for URL encoding.
+# This is to show that the URL encoding is done correctly - not all URL encoding encodes
+# the same way. We must always test the ' ' and '/' characters specifically, in addition
+# to a generic URL encoding value (e.g. $, #, etc.)
+#
+# For URL decoding, we must always test the '+' character speicifically, in addition to
+# a generic URL encoding value (e.g. $, #, etc.)
+#
+# PLEASE DO THESE TESTS FOR EVEN CASES WHERE THOSE CHARACTERS SHOULD NOT OCCUR, FOR SAFETY.
+
+
+@pytest.mark.describe(".get_c2d_topic_for_subscribe()")
+class TestGetC2DTopicForSubscribe(object):
+    @pytest.mark.it("Returns the topic for subscribing to C2D messages from IoTHub")
+    def test_returns_topic(self):
+        device_id = "my_device"
+        expected_topic = "devices/my_device/messages/devicebound/#"
+        topic = mqtt_topic_iothub.get_c2d_topic_for_subscribe(device_id)
+        assert topic == expected_topic
+
+    @pytest.mark.it("URL encodes the device_id when generating the topic")
+    @pytest.mark.parametrize(
+        "device_id, expected_topic",
+        [
+            pytest.param(
+                "my$device", "devices/my%24device/messages/devicebound/#", id="id contains '$'"
+            ),
+            pytest.param(
+                "my device", "devices/my%20device/messages/devicebound/#", id="id contains ' '"
+            ),
+            pytest.param(
+                "my/device", "devices/my%2Fdevice/messages/devicebound/#", id="id contains '/'"
+            ),
+        ],
+    )
+    def test_url_encoding(self, device_id, expected_topic):
+        topic = mqtt_topic_iothub.get_c2d_topic_for_subscribe(device_id)
+        assert topic == expected_topic
+
+
+@pytest.mark.describe(".get_input_topic_for_subscribe()")
+class TestGetInputTopicForSubscribe(object):
+    @pytest.mark.it("Returns the topic for subscribing to Input messages from IoTHub")
+    def test_returns_topic(self):
+        device_id = "my_device"
+        module_id = "my_module"
+        expected_topic = "devices/my_device/modules/my_module/inputs/#"
+        topic = mqtt_topic_iothub.get_input_topic_for_subscribe(device_id, module_id)
+        assert topic == expected_topic
+
+    @pytest.mark.it("URL encodes the device_id and module_id when generating the topic")
+    @pytest.mark.parametrize(
```

Comment:
>parametrize [](start = 17, length = 11)

If we're testing + on decode, shouldn't we be testing + on encode also?  Here an everywhere,